Mixed-source cflags

scheibelp opened the pull request:
This implements part 1 of https://github.com/spack/spack/issues/38664#issuecomment-2015660668: `.satisfies` for compiler flags is now just based on set membership.

Now that flags from multiple sources can be merged, this PR added more testing for that, and found some flag order consistency issues, which were also fixed:

1. Flags from compilers.yaml and the command line were always intra and inter source order consistent
2. Flags from dependents and packages.yaml (introduced via `require:`) were not

Before this PR, [1] was handled by going back to the original source: e.g. flags were retrieved directly from the command line spec rather than the solver.

Doing the same for [2] would be difficult, so I updated `concretize.lp` to "carry along" the associated group of flags that a flag came with (i.e. if a user asked for `zlib cflags="-a -b"`, then the solver fact for `-a` carries `-a -b` along with it, which can then be retrieved after the solver completes).

TODOs:

- [x] The solver already has mechanisms for ensuring a fixed order for flags as aggregated by all sources, but this has not been tested yet. Now that this PR allows for the mixed result to be generated, we should enforce ordering
  - Later work (and points in this TODO list) discovered some of the initial assumptions here were not correct
- [x] (March 26 2024) now that item 1 is resolved, the test has detected some issues with ordering consistency. Note that this test fails intermittently because the flags can be reordered within a given source (inter-source ordering seems consistent though)
- [x] (March 27 2024) intra-source ordering is good as of [c1b26f2](https://github.com/spack/spack/pull/41049/commits/c1b26f29781371d6bc457d43f3732a516944de53), but if `cflags` for a package come from more than one of [(a) the command-line spec (b) `require:` in packages.yaml (c)  `depends_on` from dependents] then there are inter-source ordering problems (this isn't carried through the concretize.lp file).
  - (April 16 2024) this is resolved by [4e7595c](https://github.com/spack/spack/pull/41049/commits/4e7595c0bc55c75465fa74c7b2fd786faa9683c3): flag groups from dependents, `require:`, and the command line could be reordered WRT one another because they were all considered to come from the same source. To resolve that, this alphabetically sorts on a per-source basis.
- [ ] (April 17 2024) Add a test to enforce a failure if the same flag appears with different groups (e.g. `spack spec zlib cflags="-a -b"` and the config has `packages:zlib:require:'cflags="-b -c"'`
  - (April 18 2024) I've made this very easy to detect, but haven't explicitly forbade it yet. Extra test cases added to `test_flag_order_and_grouping` demonstrate how this can result in strange behavior (but the tests confirm that it is consistent now).
- [x] (April 18 2024 - worth punting) I've added a test  `test_two_dependents_flag_mixing` for and confirmed we don't enforce the DAG ordering for flags that was originally described in `reorder_flags` (you can run the new test on the old logic with commit [19f5c6f](https://github.com/spack/spack/pull/41049/commits/19f5c6fe76ac04553d46f2658e19575c400c5a65)). For now alphabetical order is used to maintain consistency for flags mixed from multiple dependents.
  - (~June) Flags in `concretize.lp` now record the dependency that applies them (i.e. when those flags are introduced by `depends_on`), so `reorder_flags` can now apply topological ordering to those flags (and does)
- [ ] (April 16 2024 - worth punting) I was originally thinking we'd want to detect conflicting cflags in this PR and fail after the solver step if we detected a flag conflict. I'm less insistent on that now that consistent ordering is possible. But overall since there isn't fine grained control over this, I think it will still be important as a matter of helping users avoid frustration

scheibelp commented on lib/spack/spack/test/flag_mixing.py:
```diff
@@ -0,0 +1,158 @@
+        "-x6",
+        "-x7",
+        "-x4",
+    ]
```
This test is revealing instabilities WRT both inter and intra-source flag ordering. I suspect I'll need to update the solver (e.g. `concretize.lp`) to get the step-1 semantics we desire from https://github.com/spack/spack/issues/38664#issuecomment-2015660668.

tgamblin commented:
@alalazo can you profile the concretization timings on this?

alalazo commented:
Benchmarked on:

* **Spack:** 0.22.0.dev0 (db1c73deb7297d773495ccce5c1958f447afae3a)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

_without_ using any flag. The results are exactly the same (modulo one spike, which is spurious).

[radiuss.pr.csv](https://github.com/spack/spack/files/15175400/radiuss.pr.csv)
[radiuss.develop.csv](https://github.com/spack/spack/files/15175401/radiuss.develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/15175402/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/a9a74bd2-a2d2-4182-a926-8496004d990c)

adrienbernede commented:
@scheibelp thank you for working on this. Per @tgamblin request, I tested this in my original use case.
For the records it consists in setting the gcc toolchain in the compiler flags, while setting `-fsanitize=address` in the spec.
Doing exactly that, _in an environment_, here is what I get:

```
WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/solver/asp.py:3284 ==> [2024-05-06-03:06:38.553005] Warning: using "cmake@3.23.1" which is a deprecated version
WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/solver/asp.py:3326 ==> [2024-05-06-03:06:38.559854] UNKNOWN SYMBOL: attr("external_conditions_hold", NodeArgument(id='0', pkg='cmake'), 6)
WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/main.py:1071 ==> [2024-05-06-03:06:38.601007] UnsatisfiableCompilerFlagSpecError: --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 -fsanitize=address does not satisfy -fsanitize=address
WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/error.py:52 ==> [2024-05-06-03:06:38.601801] Error: --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 -fsanitize=address does not satisfy -fsanitize=address
Traceback (most recent call last):
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/main.py", line 1068, in main
    return _main(argv)
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/main.py", line 1021, in _main
    return finish_parse_and_run(parser, cmd_name, args, env_format_error)
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/main.py", line 1051, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/main.py", line 648, in _invoke_command
    return_val = command(parser, args)
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/cmd/spec.py", line 106, in spec
    env.concretize()
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/environment/environment.py", line 1241, in concretize
    return self._concretize_together(tests=tests)
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/environment/environment.py", line 1361, in _concretize_together
    *specs_to_concretize, tests=tests
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/concretize.py", line 741, in concretize_specs_together
    return _concretize_specs_together_new(*abstract_specs, **kwargs)
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/concretize.py", line 750, in _concretize_specs_together_new
    abstract_specs, tests=kwargs.get("tests", False), allow_deprecated=allow_deprecated
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/solver/asp.py", line 3562, in solve
    setup, specs, reuse=reusable_specs, output=output, allow_deprecated=allow_deprecated
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/solver/asp.py", line 822, in solve
    answers = builder.build_specs(spec_attrs)
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/solver/asp.py", line 3370, in build_specs
    _develop_specs_from_env(s, ev.active_environment())
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/solver/asp.py", line 3409, in _develop_specs_from_env
    spec.constrain(dev_info["spec"])
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/spec.py", line 3608, in constrain
    changed |= self.compiler_flags.constrain(other.compiler_flags)
  File "/usr/WS2/bernede1/Umpire/1843487-1714987750/spack-and-build-root/spack/lib/spack/spack/spec.py", line 849, in constrain
    " ".join(f for f in self[k]), " ".join(f for f in other[k])
spack.spec.UnsatisfiableCompilerFlagSpecError: --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 -fsanitize=address does not satisfy -fsanitize=address
```

tgamblin reviewed (changes requested):
ok I was reviewing, and didn't see @adrienbernede's comment -- @scheibelp can you address it?

scheibelp commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3393,7 +3406,8 @@ def _develop_specs_from_env(spec, env):
         spec.variants.setdefault("dev_path", spack.variant.SingleValuedVariant("dev_path", path))
-    spec.constrain(dev_info["spec"])
+
+    assert spec.satisfies(dev_info["spec"])
```
This 3 line diff addresses https://github.com/spack/spack/pull/41049#issuecomment-2096014945:

This uncovers a discrepancy with `constrain`, where it's unwilling to add flags (for `x.constrain(y)`, the implementation of `constrain` explicitly forbids that `x` contains flags not in `y`). I decided that WRT this PR the solution is that we shouldn't be applying more constraints after the solver completes, and skipped applying that function.

In the process of reproducing that issue (`test_dev_mix_flags`) I also found that propagation would generally not mix with other forms of flags because of a rule in the concretizer, I'll add a separate comment for that.

alalazo replied:
Not sure I entirely follow the comment above :thinking:

scheibelp replied:
* Flags should not be added at this stage (the concretizer should have combined them already)
* Therefore `constrain` is not the right method to use here
* By removing it, I sidestepped an issue where`.constrain` is not conceptually synced with the changes to `.satisfies` in this PR: for `x.constrain(y)`, the current definition of `FlagMap.constrain` (which is unaltered in this PR) will fail if x contains a flag that is not in y
* I figured this should be addressed, but I didn't want to handle that in this PR

Is that more clear?

alalazo replied:
I agree with the first couple of points, less with leaving `constrain` in an inconsistent state. I can help with that, in case.

scheibelp replied:
Should be resolved as of 3d75fbbe43b772566aeb9b2ec8ccd366c1ccacef/https://github.com/spack/spack/pull/41049#discussion_r1724045046

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1245,7 +1245,9 @@ error(100, "Compiler {1}@{2} requested for {0} cannot be found. Set install_miss
-     not attr("node_flag_set", DependencyNode, FlagType, _),
+     % Disable to pass propagate_and_spec_dep, propagate_and_pkg_dep
+     % and propagate_and_require
+     % not attr("node_flag_set", DependencyNode, FlagType, _, _),
```
This follow-up to https://github.com/spack/spack/pull/41049/files#r1591609679 is required to make all `test_propagate_and_*` tests pass except for `propagate_and_compiler_cfg`. In that sense, it isn't needed for https://github.com/spack/spack/pull/41049#issuecomment-2096014945, but I think this relaxation is sensible (I ran all the tests that made sense to me locally and am thinking the CI full run will confirm whether there are any flaws with this).

alalazo replied:
I don't have opinions on this one, but I think it changes previous semantic. Before, setting a flag, would stop propagation through that node.

scheibelp replied:
> Before, setting a flag,

By "before" I don't think you mean before this comment: but rather commits that were made by you after this comment, but that landed before this PR merged. More discussion at https://github.com/spack/spack/pull/41049#discussion_r1674822909 - we did not enforce this until after May 6 2024

scheibelp replied:
Resolved in https://github.com/spack/spack/pull/41049#discussion_r1732076592

alalazo commented:
@scheibelp Adding a note, in case the following was missed. We have known bugs in compiler flag propagation. There's a Pr https://github.com/spack/spack/pull/38478 but that seems incomplete, see https://github.com/spack/spack/pull/38478#issuecomment-2090481765

scheibelp commented:
(rebasing tells me test failures are from 3579cf0935b9e4021a212410d9d1e06d034baa49 - so I think I just need to figure out how to update my fake compiler)

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3319,8 +3321,8 @@ def node_compiler_version(self, node, compiler, version):
+    def node_flag(self, node, flag_type, flag, flag_group):
+        self._specs[node].compiler_flags.add_flag(flag_type, flag, False, flag_group)
 
     def node_flag_source(self, node, flag_type, source):
         self._flag_sources[(node, flag_type)].add(source)
```
This is no longer used, but see below -- I think it should be.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3367,15 +3369,19 @@ def virtual_on_edge(self, parent_node, provider_node, virtual):
+        imposes order afterwards. The order is:
+
+        1. Compiler flags should always come first
+        2. Command-line flags should always come last
```
this should probably be the last item in the list (i.e. this should be number 4).  I had to read the comment a few times and it was confusing to me that this was item 2.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3367,15 +3369,19 @@ def virtual_on_edge(self, parent_node, provider_node, virtual):
+           flags from that source should retain their order and grouping, e.g.
+           for `y cflags="-z -a"` "-z" and "-a" should never have any intervening
+           flags inserted, and should always appear in that order.
+        4. Flags from requirements, and dependents should be grouped and ordered
+           alphabetically (on a per-source basis, accounting for the prior point).
```
@scheibelp why does this order dependent flags alphabetically when the prior logic used DAG order?  DAG order ensures that you get a deterministic ordering, but it also ensures that flags on a node override those inherited from its dependents. If you sort alphabetically you're not ensuring that.... it can change which node overrides which depending on the flags provided.  e.g.:

```
     A
    / \
   B   C
    \ /
     D
```

Any flags on D inherited from B and C should come after those inherited by A -- the order shouldn't depend on the name of the particular flag.  e.g., you don't want `-O3` to always take precedence over `-O2` regardless of whether A, B, or C specified it.

Can't we both preserve DAG order *and* preserve source grouping? I think you can add the `flag_group` to `node_flag_source` and preserve the `.index()`-based sorting I had below, but also preserve flag groups.

scheibelp replied:
I point out in the PR description that while we thought we did this, we don't. You can check out [19f5c6f](https://github.com/spack/spack/pull/41049/commits/19f5c6fe76ac04553d46f2658e19575c400c5a65) to test that yourself (new DAG ordering test with old logic).

tgamblin replied:
are you referring to `test_flag_order_and_grouping`?  It doesn't look like it's trying to enforce DAG order... which test do you want me to run on that?

scheibelp replied:
`test_two_dependents_flag_mixing`: this test only checks flag ordering for flags introduced by `depends_on` vs. any other form of mixing. If you run it 5-10 times it will fail some and succeed some.

scheibelp replied:
I removed `node_flag_source` and introduced `source` as a property of the `node_flag` attribute to properly associate introduced flags with their "origin" (i.e. whether they originated from a `depends_on`, or from a `require:`.

Right now, `depends_on` is the only origin that can be applied to one package from multiple sources; `require:` might do that in the future, so I added a `ConstraintOrigin` class. That could be removed while supporting everything we do now. Either way, the notion of origin needs to be plumbed through relevant methods in `asp.py` (e.g. for `package_dependencies_rules() -> condition() -> _get_condition_id() -> spec_clauses()`, we need to pass the origin info from the first to the last element, through all intermediates).

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3367,15 +3369,19 @@ def virtual_on_edge(self, parent_node, provider_node, virtual):
 
         The solver determines which flags are on nodes; this routine
-        imposes order afterwards.
+        imposes order afterwards. The order is:
```
One concern I have with this is that we aren't accounting for the override precedence that was added in the `cc` wrapper in https://github.com/spack/spack/pull/37376.  I *think* what we will probably want at some point is not a strict first/last rule, but a "this flag overrides that one" rule.

The position to put a flag in to force it to override depends on the flag type, and:
1. you may have multiple types of flags (e.g. both `-l` and `-L`/`-I` flags) in the same group, and
2. you can't split a flag group apart b/c you want to preserve internal ordering

I think we'd need to do some surgery on flag groups to do this right, but that is beyond the scope here. I *do* think we should at least preserve DAG precedence (i.e. children come after parents on the CLI) -- see below -- since that  behavior was present before.

scheibelp replied:
Presumably the severable part of this you want done here is addressed by https://github.com/spack/spack/pull/41049#discussion_r1593099877

adrienbernede commented:
@scheibelp the current state of the branch does combine flags successfully in my use case.

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -786,6 +786,8 @@ class CompilerFlag(str):
     def __new__(cls, value, **kwargs):
         obj = str.__new__(cls, value)
         obj.propagate = kwargs.pop("propagate", False)
+        obj.flag_group = kwargs.pop("flag_group", value)
+        obj.source = kwargs.pop("source", None)
```
Can we extend the docstring to explain what `flag_group` and `source` are meant to be? Also why are we using `kwargs`? Can we use explicit arguments?

scheibelp replied:
Yes, but for the meantime:

* `flag_group` is the group of flags that introduced a given flag: e.g. if you say `cflags=-x -y`, there will be a `-x` flag tracked in the solver, and we want to remember that it came with `-y`, so the `flag_group` keeps track of `-x -y`
* `source` tracks the type of constraint that introduced the flag: it is needed because the same package can introduce flags via multiple mechanisms (so in terms of applying a consistent order, we have to track this info and assign an arbitrary order based on where the flag constraint came from).

alalazo commented on lib/spack/spack/solver/display.lp:
```diff
@@ -13,6 +13,8 @@
 #show attr/2.
 #show attr/3.
 #show attr/4.
+#show attr/5.
+#show attr/6.
```
Having `attr/5` and `attr/6` (long list of args) probably means that we can structure the knowledge representation better. Maybe we should try to group things in nested facts?

scheibelp replied:
We already had `attr/5`, but for whatever reason, we didn't need it in the solution.

That being said, `FlagType, Flag, FlagGroup` always appear together, so could potentially be consolidated, if I understand what you mean by "nested fact" (I think `node_flag` would be an example of that?).

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1010,23 +997,24 @@ variant_is_propagated(PackageNode, Variant) :-
+  propagate(node(PackageID, Package), node_flag(FlagType, Flag, FlagGroup, Source), _),
+  % Disable to pass propagate_and_spec_dep, propagate_and_pkg_dep
+  % and propagate_and_require
+  % not attr("node_flag_set", node(PackageID, Package), FlagType, _, _, _),
```
(the reincarnation of https://github.com/spack/spack/pull/41049#discussion_r1591611535)

scheibelp replied:
Resolved in https://github.com/spack/spack/pull/41049#discussion_r1732076592

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1010,23 +997,24 @@ variant_is_propagated(PackageNode, Variant) :-
   node_compiler(node(PackageID, Package), CompilerID),
   node_compiler(SourceNode, CompilerID),
-  attr("propagate", SourceNode, node_flag(FlagType, Flag), _),
+  attr("propagate", SourceNode, node_flag(FlagType, Flag, FlagGroup, Source), _),
```
(question) I found it interesting that this was expressed as an `attr` vs. as:

`propagate(SourceNode, node_flag(FlagType, Flag, FlagGroup, Source), _)`

I see that:

* `asp.py` generates `propagate` as an attribute like `attr(propagate...)`
* `concretize.lp` generates a `propagate()` predicate from this
* All other rules in `concretize.lp` interact with this information via `propagate()` vs. `attr(propagate...)`

Why not use `propagate()` here?

Also, why doesn't `propagate()` encode edge info? As written, this seems susceptible to undesired associations:

```
  v
 / \ 
w   y  # ^w cflags==g, ^y cflags==g
|   |
x   z

Say x, y, z have the same compiler and w has a different compiler:
as encoded, flag propagation info can "jump" across branches.
```

alalazo replied:
Only the sources of the propagation have the:
```
attr("propagate", Node, PropagatedAttribute)
```
Internally the propagation uses `propagate` facts.

scheibelp replied:
* I assume your response is to the first part of my comment (i.e., it is not addressing the concern about the lack of edge info in `propagate()`)
* Even so, it does not conflict with what I say: this line could use `propagate()` rather than `attr("propagate"...` (and IMO it would be worthwhile because it initially puzzled me why it differed from all the other references)

alalazo replied:
The `propagate` rules just move some facts down the DAG. They're simple rule and propagate "dumbly" so that complexity is kept O(edges). The propagated facts should contain enough information for a node to decide locally whether a fact in `propagate` should have any effect on its `attr` or not.

`attr("propagate", ...)` is currently used to kick-off propagation, so it's only on the roots. We can use any other mechanism to kick-off `propagate` rules, but why?

scheibelp replied:
> We can use any other mechanism to kick-off propagate rules, but why?

I'm not against the mechanism for kicking it off

The general logic for dependencies like `X -> Y -> Z` is

* `attr("propagate")` on X generates `propagate()` on X
* `propagate()` on X applies transitively to Y/Z

Everything in this lineage (including the root) has `propagate()`, so I wondered why this constraint is conditioning on `attr("propagate")`.

Does it not seem strange that we have all of `propagate(X/Y/Z)`, but that we effectively ask "if Z has `propagate()`, and if we initialized propagation on X..."?

One reason I can think of *not* to swap to `propagate()` is that this may lead to more comparisons (because, as you pointed out, `attr("propagate"...` is just on the roots). Although that's not clear from an initial reading.

Overall:

* I understand how this works
* I thought it was confusing
* I wondered if you thought it was too
* It's not critical IMO to replace `propagate()` with `attr("propagate")`
* However, whether you do or don't, there is a potential issue for "branch jumping" (described in detail in 1st comment)

scheibelp commented on lib/spack/spack/test/concretize.py:
```diff
@@ -437,11 +437,11 @@ def test_compiler_flags_differ_identical_compilers(self, mutable_config, clang12
                 "hypre cflags=='-g' ^openblas cflags='-O3'",
                 ["hypre cflags='-g'", "^openblas cflags='-O3'"],
-                ["^openblas cflags='-g'"],
+                [],
```
There's a minor tension here:

1. I see the appeal of reusing the `*flags=...` syntax to override propagation
  * OTOH how do we, for `x->y`, propagate flags from x to y while also adding more flags for y? I'm inclined to say we need more syntax to address this
2. Currently, the `.lp` constructs are not sufficiently fine-grained to distinguish between this and other cases checked by the `test_propagate_and_...` tests added by this PR
  * e.g. for this test to pass, `depends_on(foo cflags=...)` would disable propagation

I'm guessing the desired way to address this is to figure out how to keep track of this in `concretize.lp`: essentially node flag settings from input specs should disable propagated flags (but when flags are set by other means, propagation should be allowed). This would allow both tests to pass. I wanted to make sure this was the desired approach in case enforcing this was an afterthought, and we decide we want to support this behavior a different way.

Other ways to resolve this:

* Make `test_propagate_and_*` xfail for now (effectively, flags set via any other mechanism will disable propagation)
* Leave this (e.g. if we weren't advertising it or depending on it anywhere, don't give users a mechanism to override propagation for now)

scheibelp replied:
@becker33 after speaking with @alalazo, he mentioned these new enforcements came from a discussion with you about the intended behavior of 

```
X cflags==-g
|
Y cflags=-O3   (this is supposed to override -g, not add to it)
|
Z
```

are you particularly attached to this behavior? The current syntax must rule out either override or append; I would have assumed appending would be a more common desire.

scheibelp replied:
This is restored in [f7774eb](https://github.com/spack/spack/pull/41049/commits/f7774ebe71409ec474be0994dacf4728f22d3398)

scheibelp committed:
Mixed-source cflags

This is the first step in improving compiler flag semantics in Spack. It implements
set membership semantics for compiler flags in `Spec.satisifes()`, so that:

* one spec can satisfy another *if it contains a subset of the other's compiler flags*, and
* constraining a spec's compiler flags simply adds flags.

This enables flags from multiple sources to be merged and adds testing for merging flags.

Some flag order consistency issues are also fixed:

1. Flags from `compilers.yaml` and the command line were always intra- and inter-source
   order consistent.
2. Flags from dependents and packages.yaml (introduced via `require:`) were not.

(1) is handled by going back to the original source, i.e., flags are retrieved directly
from the command line spec rather than the solver.

(2) is handled by enforcing topological ordering when multiple dependents impose flags on
a dependency.  e.g., for this graph:

```
   a
  /|\
 b | c
  \|/
   d
```

If `a`, `b`, and `c` impose flags on `d`, the combined flags on `d` will contain the
flags of `a`, `b`, and `c` -- in that order. This may need to change later (e.g. we may
need to reverse the order for certain types of flags based on the way they take
precedence on the command line), but we have gone with this comparatively simple option
for the first cut.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

alalazo commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -1465,7 +1465,7 @@ def test_abstract_contains_semantic(lhs, rhs, expected, mock_packages):
         # Note: the semantic is that of sv variants, not mv variants
         (Spec, "cppflags=-foo", "cppflags=-bar", (False, False, False)),
-        (Spec, "cppflags='-bar -foo'", "cppflags=-bar", (False, False, False)),
+        (Spec, "cppflags='-bar -foo'", "cppflags=-bar", (False, True, False)),
```
(Test with wrong semantic) This test seems logically wrong. If one of the two specs satisfies the other ($\subseteq$ semantic) they must intersect ($\cap$ semantic):
```suggestion
        (Spec, "cppflags='-bar -foo'", "cppflags=-bar", (True, True, False)),
```

scheibelp replied:
I have to change the implementation of intersects for this to work.

Right now (i.e. outside of this PR), it is possible for there to be satisfaction without intersection: that is because `intersects` examines `.propagate`, and `satisfies` does not.

With that in mind I want to make sure we agree on the desired semantics:

* Ignoring `.propagate` I think the definition of `intersects` for a `FlagMap` should just be `return True` (because the point of this PR is to allow any combination of flags, including `-O2` and `-O3`, for example).
* I'm not clear why propagation is considered as a part of intersection, so I still think `.intersects` should always return `True`:
  * Why should `cflags==-O2` not intersect with `cflags=-O2`? My understanding is that `intersects` should only return `False` if the specs conflict with each other. Not only do these not conflict, I think they satisfy each other 
    * Or at least, `==-O2` satisfies `=-O2` (I assume this is not controversial)
    * I'd argue the other direction too:  `=-O2` satisfies `==-O2`, because in terms of the immediate spec, the result is the same (in a reuse circumstance for example, this wouldn't matter because regardless of propagation, satisfaction would be resolved via the dependencies).
  * This was introduced in https://github.com/spack/spack/commit/bc209c470d07621ba9af277dab71d048d11f2844 and continued in https://github.com/spack/spack/commit/d54611af2c56a04ed0990055d357b30b54b30a5d (the commit responsible for the creation of `intersects`, written by you, so I think you're the person to litigate this with)

alalazo replied:
In https://github.com/spack/spack/commit/d54611af2c56a04ed0990055d357b30b54b30a5d I just split the function trying to keep previous semantic. I also think propagation should not enter in `satisfies` / `intersects`  semantic.

scheibelp replied:
OK, `.intersects` now always returns `True`, and I have done my best to also update the accompanying `.constrain` function to achieve consistency (see new test `test_constrain_compiler_flags`).

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1025,6 +1025,39 @@ def __iter__(self):
+    result.
+    """
+
+    _src_id_suffix = [("depends_on", "_dep"), ("require", "_req")]
```
We should enumerate flag origins, and avoid strings. This should help _not_ having to deal with typos and improper source types.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1439,7 +1475,11 @@ def condition(
+                cache=self._trigger_cache,
+                body=True,
+                transform=transform_required,
+                source=source,
```
Instead of adding arguments down the chain for this specific purpose, why we don't use the transformation functions?[^1]

[^1]: The `source` argument makes sense only in a few places. Those places can define a transformation function and avoid changes to the interface.

scheibelp replied:
The `transform_*` functions add/remove clauses vs. modifying them (`_spec_clauses` needs to attach the `source` to generated `node_flag` attributes); overall, I would consider it confusing to bundle any of the currently generated clauses related to node flags into a callback.

alalazo replied:
> The transform_* functions add/remove clauses vs. modifying them

Who says that? Also, what's the difference? :slightly_smiling_face:  The `transform_` function takes a list of facts, does what it needs to do, and returns the final list. I added that callback to avoid having a long list of arguments, each one used in a very specific place.

alalazo reviewed (changes requested):
Combining flags works as expected when tested on a few simple use cases. 

I left a few comments on the implementation. Also, one unit test has the wrong semantic (and that's why I'm requesting changes, not just leaving comments). 

In general, I think we can simplify the encoding - I might take some time this week to check if I can come up with some more detailed suggestion.

scheibelp committed:
expand CompilerFlag docstring

scheibelp committed:
replace string-based identifiers with an enum

scheibelp committed:
auto style fix

scheibelp committed:
flagsets always intersect

scheibelp committed:
partial update of tests WRT new flagmap.inersects semantics

scheibelp committed:
demonstrative update to flagmap.constrain

scheibelp committed:
auto style fix

scheibelp committed:
(update example tests alongside flagmap.constrain)

scheibelp committed:
bad calculation of changed for flagmap.constrain

scheibelp committed:
test for asymmetric constrain on flagmaps

scheibelp committed:
auto style fix

scheibelp committed:
check that .constrain preserves propagation semantics

scheibelp committed:
auto style fix

scheibelp committed:
more explicit propagation check

scheibelp committed:
auto style fix

scheibelp committed:
update test docstring w/ more explanation

scheibelp committed:
expand comment about inconsistency

scheibelp committed:
consistent propagation for FlagMap.constrain

scheibelp committed:
demonstrate interaction between flag group mixing and propagation with .constrain

scheibelp commented on lib/spack/spack/spec.py:
```diff
@@ -800,52 +832,38 @@ def __init__(self, spec):
+                        y.propagate = True
+
+        # TODO: what happens if flag groups with a partial (but not complete)
+        # intersection specify different behaviors for flag propagation?
```
`FlagMap.constrain` has been updated to mostly account for https://github.com/spack/spack/pull/41049#discussion_r1723532851, but I can't claim to have exhausted all possible interactions between propagation, `.constrain()`, and the behavior of the ASP-based-concretization.

alalazo commented on lib/spack/spack/test/flag_mixing.py:
```diff
@@ -0,0 +1,327 @@
+    assert result["y"].satisfies("cflags='-fsanitize=address -f1'")
+
+
+def test_diamond_dep_flag_mixing(concretize_scope, test_repo):
+    root_spec1 = Spec("t").concretized()
+    spec1 = root_spec1["y"]
+    assert spec1.satisfies('cflags="-c1 -c2 -d1 -d2 -e1 -e2"')
+    assert spec1.compiler_flags["cflags"] == "-c1 -c2 -e1 -e2 -d1 -d2".split()
```
For this test (are there others here?) we need to write somewhere that the result being asserted depends on the implementation details of our topo-sorting.

scheibelp replied:
I added a docstring explaining this.

alalazo reviewed (changes requested):
As we discussed, a few modifications that don't touch the underlying mixing logic in https://github.com/scheibelp/spack/pull/39

scheibelp committed:
partial application of diff from suggestion

scheibelp committed:
restore original behavior where cmdline spec flags override propagation from dependents

scheibelp committed:
auto style fix

scheibelp committed:
test propagation on dependency overriding propagation from parent

scheibelp committed:
explain that test behavior relies on implementation details of .traverse

scheibelp committed:
replace single 'source' string arg with a Context object that could hold additional state

scheibelp committed:
aggregate transform_required/transform_imposed into Context

scheibelp committed:
merge ConstraintOriginType into ConstraintOrigin (a separate class is not needed to define the enum values)

scheibelp committed:
remove test that is now covered by flag_mixing test module (and is also now working in this PR)

scheibelp commented:
@alalazo recent updates:

- Allow command-line `cflags=` override of propagation while allowing other flag sources (like `require:` and `depends_on`) to add to propagation
- `FlagMap.intersects` and `FlagMap.constrain` are now consistent with new meaning of `FlagMap.satisfies`
- Most changes from your proposed branch are merged with one exception: the `transform_`-based modification of `spec_clauses` to modify flag source:
    - In general I'm concerned about the complexity of and approach that generates something wrong and then make it right after the fact
    - I replaced that with an approach that bundles the transform_ functions together with a `source` into a `*Context` object
        - `spec_clauses` (lowest level) needs a `SourceContext`
        - `_get_condition_id` needs a `ConditionIdContext` (which is also a `SourceContext`)
        - `condition()` (highest level) needs a `ConditionContext` (which is also a `SourceContext`)
        - Overall: this retains the `transform_*` functions where they are originally used, and achieves the appropriate pass-through of `source` without increasing the number of parameters (other than in `spec_clauses`)
    - I'm curious what you think about the approach at a high level: does it seem like an improvement? If not, I can take another crack at refactoring. 

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1411,8 +1476,7 @@ def condition(
         name: Optional[str] = None,
         msg: Optional[str] = None,
-        transform_required: Optional[TransformFunction] = None,
-        transform_imposed: Optional[TransformFunction] = remove_node,
```
I still need to double check that we maintained the same default everywhere.

scheibelp replied:
I was not doing that (although I'll note that no tests failed because of that): the latest updates restore this and also tried to make that clear (I now instantiate `ConditionContext` and set `.transform_imposed` on it in this function if it is `None`).

alalazo replied:
> I was not doing that (although I'll note that no tests failed because of that):

That could be: `transform_imposed` is not used if `imposed_spec` is `None`.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1411,8 +1476,7 @@ def condition(
         msg: Optional[str] = None,
-        transform_required: Optional[TransformFunction] = None,
-        transform_imposed: Optional[TransformFunction] = remove_node,
+        context: Optional[ConditionContext] = None,
```
The docstring needs an update in this function

scheibelp replied:
I've updated docstrings (and also added some to the `*Context` objects.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1796,6 +1870,7 @@ def spec_clauses(
         expand_hashes: bool = False,
         concrete_build_deps=False,
         required_from: Optional[str] = None,
+        context: Optional[SourceContext] = None,
```
Here we miss many updates to the docstring (not your fault), but adding `context` would be appreciated.

alalazo reviewed (commented):
The context approach is fine with me. It can probably be simplified a bit, and we might find better names for classes, but we can leave it for a future PR.

There is still some minor clean-up needed.

scheibelp committed:
merge from develop (minor conflicts with 46176)

scheibelp committed:
some docs

scheibelp committed:
more docs

scheibelp committed:
had to move conditionidcontext definition; more docs; initialize transform_imposed in a manner that matches behavior prior to this PR

scheibelp committed:
auto style fix

scheibelp committed:
more docs

alalazo reviewed (approved):
This looks good to me. We can improve the code further in a following PR.
